README
------------------------------------------------
Freeverb3 Impulse Response Processor (freeverb3)
------------------------------------------------

Freeverb3 is a signal processing library.

This library includes many types of audio processing effects and signal
generators. The linear feedback shift register (LFSR) maximum length
sequence (MLS) generator, multiple types of high quality mono and stereo reverb
effects, an multi slot impulse response processor, multiband hardknee and softknee
compressor and limiter, a stereo enhancer, a limiter and a lot of
types of infinite impulse respose (IIR) and finite impulse response (FIR)
filters are available.

The convolution reverb algorithm is highly optimized
for x86 processors by using hand optimized codes of 3DNow!/SSE/SSE3/AVX/FMA/FMA4
(single precision), SSE2/SSE4/AVX/FMA/FMA4 (double precision), which can be
used for zero latency realtime sound effect processing. This library supports
any environment which GCC supports including x86/ppc/sparc. 

This library includes audacious effect plugins and sample programs,
which can be used to test the sound effect easily.
Please report any bugs or advices.

The hand-optimized SIMD codes for some CPUs are included in this library,
which will provide fast convolution processing.

Dependencies of libfreeverb3
============================

* FFTW3 [http://www.fftw.org/]
  This library needs at least one precision version
  (single or double or long double) of fftw3 libraries,
  which are called fftw3f or fftw3 or fftw3l.
  fftw-3.1.2 was tested.

Dependencies of programs
========================

Sample programs      : libsndfile
Jack sample programs : libsndfile jack
                       (The configdb subset library is included in jack dir,
		       so jack programs do not depend on audacious.)
audacious            : libsndfile audacious glib2 GTK+3 gthread pango

* libsndfile [http://www.mega-nerd.com/libsndfile/]
  libsndfile-1.0.17 was tested.

* jack [http://jackaudio.org/]
  jackd version 0.101.1 was tested.

* audacious [https://audacious-media-player.org/]
  audacious >= 3.5 is required.

* GTK+ [http://www.gtk.org/]
  gtk-3.x >= 3.0.0 is required.

* glib2 [http://www.gtk.org/]
  glib-2.x >= 2.6.0 is required.

* gthread, pango

If you are building this from the source repository sources,
you will need the following tools:

* autotools (autoconf, automake and libtool)
* autogen

configure OPTIONS
=================

See configure --help.

SIMD optimization codes for the impulse convolution
===================================================

binutils >= 2.19.51.0.12 is required to build SIMD codes including AMD FMA4.

3DNow! uses primitive 3DNow! codes.
This can be used on AMD CPUs which have 3DNow! instructions.
(K6-2, Athlon, ...)

SSE can be used on Intel or AMD CPUs which have SSE instrunctions.
(PentiumIII, ..., AthlonXP, Athlon64, ...)
SSE is used on x86_64/amd64.

SSE2 can be used on Intel or AMD CPUs which have SSE2 instrunctions.
(Pentium4, ..., Opteron, ...)
SSE2 is enabled on x86_64/amd64.

Use --disable-x86simd to disable x86 SIMD codes on x86/x86_64/amd64.

Supported instructions:
single floating point | 3DNow!, SSE, SSE3, AVX, FMA3, FMA4
double floating ponit | SSE2, SSE4, AVX, FMA3, FMA4

CONTACTS
========

Freeverb3 is written by Teru Kamogashira.
The Freeverb3 home page is at:

	http://www.nongnu.org/freeverb3/

COPYING
=======

Copyright:  Copyright (C) 2006-2018 Teru Kamogashira
Licence:    GPL v2

ChangeLog	This is a current ChangeLog file.
ChangeLog.v2	This is a ChangeLog of obsolete versions.
samples/IR/*.wav

Copyright:  Copyright (C) 2006-2018 Teru Kamogashira
Licence:    GPL v2

This file is part of Freeverb3.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.




ChangeLog
Sat Mar 24 19:55:32 2018 +0900 teru <teru@debian57.cc.h.u-tokyo.ac.jp> [dca8bcc] MLS default table fix
Sun Feb 4 15:13:51 2018 +0900 Teru Kamogashira <tkamogashira@users.osdn.me> [eff88d5] Jack build updates
Sun Feb 4 15:08:01 2018 +0900 Teru Kamogashira <tkamogashira@users.osdn.me> [0cbed50] Jack build updates
Sun Feb 4 14:17:10 2018 +0900 Teru Kamogashira <tkamogashira@users.osdn.me> [e87a9cd] Audacious-3.9 updates
Sun Jul 16 16:59:44 2017 +0900 teru <teru@debian57.cc.h.u-tokyo.ac.jp> [63bb2d0] oxsave fix
Tue Nov 10 15:16:40 2015 +0900 Teru Kamogashira <tkamogashira@users.sourceforge.jp> [2f0aa92] Filename null
Mon Nov 2 19:51:50 2015 +0900 Teru Kamogashira <tkamogashira@users.sourceforge.jp> [796b552] jack compilation fix
Mon Nov 2 18:56:25 2015 +0900 Teru Kamogashira <tkamogashira@users.sourceforge.jp> [2cc6d2f] Added so36.cpp and fixed jack compilation errors.
Sun Nov 1 19:26:35 2015 +0900 Teru Kamogashira <tkamogashira@users.sourceforge.jp> [eeb3538] Audacious 3.6 updates / 3.5 support fix
Tue Oct 27 12:48:09 2015 +0900 Teru Kamogashira <tkamogashira@users.sourceforge.jp> [71d58b9] Audacious 3.6 API updates
Tue Oct 27 11:41:04 2015 +0900 Teru Kamogashira <tkamogashira@users.sourceforge.jp> [78c5205] Audacious 3.6 API updates
Sat Jun 7 22:53:00 2014 +0900 Teru Kamogashira <teru@sodan.ecc.u-tokyo.ac.jp> [609e959] audacious-3.5 updates.
Sat Jun 7 22:37:09 2014 +0900 Teru Kamogashira <teru@sodan.ecc.u-tokyo.ac.jp> [09cd2c4] audacious-3.5 updates
Sat Jun 7 22:21:41 2014 +0900 Teru Kamogashira <teru@sodan.ecc.u-tokyo.ac.jp> [9f60ae8] fixed cpuid segfault on x64.
Tue Mar 11 20:05:27 2014 +0900 Teru Kamogashira <teru@sodan.ecc.u-tokyo.ac.jp> [bc9e858] Copyright notices were changed.
Thu Dec 19 19:34:42 2013 +0900 Teru Kamogashira <teru@sodan.ecc.u-tokyo.ac.jp> [80dd318] Copyright notices were fixed.
Thu Dec 19 19:02:08 2013 +0900 Teru Kamogashira <teru@debian> [c80cacf] Cleaning up unused files.
Thu Dec 19 18:53:03 2013 +0900 Teru Kamogashira <teru@debian> [b72e8ea] Deleting unused entries.
Wed Dec 18 20:30:15 2013 +0900 Teru Kamogashira <teru@debian> [85ba48a] Initial Import


freeverb/mls.cpp
/**
 *  Maximum Length Sequence generator based on Linear Feedback Shift Register
 *
 *  Copyright (C) 2006-2014 Teru Kamogashira
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include "freeverb/mls.hpp"
#include "freeverb/fv3_type_float.h"
#include "freeverb/fv3_ns_start.h"

// Reference: XILINX Linear Feedback Shift Register v3.0 DS257 (v1.0) March 28, 2003
// Polynominal 1-Terms(Xn)
const uint64_t FV3_(lfsr)::LFSR_Tap_Poly[FV3_MLS_MAX_BITS+1][FV3_MLS_MAX_TAPS] =
{{0,0,0,0,0,0,},
 {1,0,0,0,0,0,}, // 1 1+X
 {1,2,0,0,0,0,}, // 2 1+X+X^2
 {1,3,0,0,0,0,}, // 3 1+X+X^3
 {4,1,0,0,0,0,},
 {5,2,0,0,0,0,},
 {6,1,0,0,0,0,},
 {7,1,0,0,0,0,},
 {8,2,3,4,0,0,},
 {9,4,0,0,0,0,},
 {10,3,0,0,0,0,},
 {11,2,0,0,0,0,},
 {12,6,8,11,0,0,},
 {13,9,10,12,0,0,},
 {14,9,11,13,0,0,},
 {15,1,0,0,0,0,},
 {16,1,3,12,0,0,},
 {17,3,0,0,0,0,},
 {18,7,0,0,0,0,},
 {19,13,17,18,0,0,},
 {20,3,0,0,0,0,},
 {21,2,0,0,0,0,},
 {22,1,0,0,0,0,},
 {23,5,0,0,0,0,},
 {24,1,2,7,0,0,},
 {25,3,0,0,0,0,},
 {26,20,24,25,0,0,},
 {27,22,25,26,0,0,},
 {28,3,0,0,0,0,},
 {29,2,0,0,0,0,},
 {30,24,26,29,0,0,},
 {31,3,0,0,0,0,},
 {32,10,30,31,0,0,},
 {33,13,0,0,0,0,},
 {34,7,32,33,0,0,},
 {35,2,0,0,0,0,},
 {36,11,0,0,0,0,},
 {37,32,33,34,35,36,},
 {38,32,33,37,0,0,},
 {39,4,0,0,0,0,},
 {40,2,19,21,0,0,},
 {41,3,0,0,0,0,},
 {42,1,22,23,0,0,},
 {43,1,5,6,0,0,},
 {44,1,26,27,0,0,},
 {45,1,3,4,0,0,},
 {46,1,20,21,0,0,},
 {47,5,0,0,0,0,},
 {48,1,27,28,0,0,},
 {49,9,0,0,0,0,},
 {50,1,26,27,0,0,},
 {51,1,15,16,0,0,},
 {52,3,0,0,0,0,},
 {53,1,15,16,0,0,},
 {54,1,36,37,0,0,},
 {55,24,0,0,0,0,},
 {56,1,21,22,0,0,},
 {57,7,0,0,0,0,},
 {58,19,0,0,0,0,},
 {59,1,21,22,0,0,},
 {60,1,0,0,0,0,},
 {61,1,15,16,0,0,},
 {62,1,56,57,0,0,},
 {63,1,0,0,0,0,},
 {64,1,3,4,0,0,},
 {65,18,0,0,0,0,},
 {66,1,9,10,0,0,},
 {67,1,9,10,0,0,},
 {68,9,0,0,0,0,},
 {69,2,27,29,0,0,},
 {70,1,15,16,0,0,},
 {71,6,0,0,0,0,},
 {72,6,47,53,0,0,},
 {73,25,0,0,0,0,},
 {74,1,15,16,0,0,},
 {75,1,10,11,0,0,},
 {76,1,35,36,0,0,},
 {77,1,30,31,0,0,},
 {78,1,19,20,0,0,},
 {79,9,0,0,0,0,},
 {80,1,37,38,0,0,},
 {81,4,0,0,0,0,},
 {82,3,35,38,0,0,},
 {83,1,45,46,0,0,},
 {84,13,0,0,0,0,},
 {85,1,27,28,0,0,},
 {86,1,12,13,0,0,},
 {87,13,0,0,0,0,},
 {88,1,71,72,0,0,},
 {89,38,0,0,0,0,},
 {90,1,18,19,0,0,},
 {91,1,83,84,0,0,},
 {92,1,12,13,0,0,},
 {93,2,0,0,0,0,},
 {94,21,0,0,0,0,},
 {95,11,0,0,0,0,},
 {96,2,47,49,0,0,},
 {97,6,0,0,0,0,},
 {98,11,0,0,0,0,},
 {99,2,45,47,0,0,},
 {100,37,0,0,0,0,},
 {101,1,6,7,0,0,},
 {102,1,66,67,0,0,},
 {103,9,0,0,0,0,},
 {104,1,10,11,0,0,},
 {105,16,0,0,0,0,},
 {106,15,0,0,0,0,},
 {107,2,63,65,0,0,},
 {108,31,0,0,0,0,},
 {109,1,6,7,0,0,},
 {110,1,12,13,0,0,},
 {111,10,0,0,0,0,},
 {112,2,43,45,0,0,},
 {113,9,0,0,0,0,},
 {114,1,81,82,0,0,},
 {115,1,14,15,0,0,},
 {116,1,70,71,0,0,},
 {117,2,18,20,0,0,},
 {118,33,0,0,0,0,},
 {119,8,0,0,0,0,},
 {120,7,111,118,0,0,},
 {121,18,0,0,0,0,},
 {122,1,59,60,0,0,},
 {123,2,0,0,0,0,},
 {124,37,0,0,0,0,},
 {125,1,107,108,0,0,},
 {126,1,36,37,0,0,},
 {127,1,0,0,0,0,},
 {128,2,27,29,0,0,},
 {129,5,0,0,0,0,},
 {130,3,0,0,0,0,},
 {131,1,47,48,0,0,},
 {132,29,0,0,0,0,},
 {133,1,51,52,0,0,},
 {134,57,0,0,0,0,},
 {135,11,0,0,0,0,},
 {136,1,125,126,0,0,},
 {137,21,0,0,0,0,},
 {138,1,7,8,0,0,},
 {139,3,5,8,0,0,},
 {140,29,0,0,0,0,},
 {141,1,31,32,0,0,},
 {142,21,0,0,0,0,},
 {143,1,20,21,0,0,},
 {144,1,69,70,0,0,},
 {145,52,0,0,0,0,},
 {146,1,59,60,0,0,},
 {147,1,37,38,0,0,},
 {148,27,0,0,0,0,},
 {149,1,109,110,0,0,},
 {150,53,0,0,0,0,},
 {151,3,0,0,0,0,},
 {152,1,65,66,0,0,},
 {153,1,0,0,0,0,},
 {154,2,127,129,0,0,},
 {155,1,31,32,0,0,},
 {156,1,115,116,0,0,},
 {157,1,26,27,0,0,},
 {158,1,26,27,0,0,},
 {159,31,0,0,0,0,},
 {160,1,18,19,0,0,},
 {161,18,0,0,0,0,},
 {162,1,87,88,0,0,},
 {163,1,59,60,0,0,},
 {164,1,13,14,0,0,},
 {165,1,30,31,0,0,},
 {166,1,38,39,0,0,},
 {167,6,0,0,0,0,},
 {168,2,15,17,0,0,},};

FV3_(lfsr)::FV3_(lfsr)()
{
  setBitSize(4);
  initState();
  setDefaultTap();
}

FV3_(lfsr)::~FV3_(lfsr)()
{
  ;
}

void FV3_(lfsr)::setBitSize(long size)
{
  if(size > FV3_MLS_MAX_BITS||size <= 0) return;
  size_2 = size;
  setDefaultTap();
}

long FV3_(lfsr)::getBitSize()
{
  return size_2;
}

void FV3_(lfsr)::initState()
{
  currentLFSR = 1;
  cLFSR2 = 0;
  cLFSR3 = 0;
}

void FV3_(lfsr)::setDefaultTap()
{
  uint64_t tap = 0, tap2 = 0, tap3 = 0;
  for(long i = 0;i < FV3_MLS_MAX_TAPS;i ++)
    {
      uint64_t tbit = LFSR_Tap_Poly[size_2][i];
      if(tbit != 0)
	{
	  if(tbit <= 64)
	    {
	      tap |= (uint64_t)1 << (tbit-1);
	    }
	  else if(tbit <= 128)
	    {
	      tap2 |= (uint64_t)1 << (tbit-1-64);
	    }
	  else if(tbit <= 168)
	    {
	      tap3 |= (uint64_t)1 << (tbit-1-128);
	    }
	  else
	    {
	      // table error
	      throw std::bad_alloc();
	    }
	}
    }
  setTap(tap, tap2, tap3);
}

void FV3_(lfsr)::setTap(int64_t tap)
{
  currentTap = tap;
}

void FV3_(lfsr)::setTap(int64_t tap, int64_t tap2, int64_t tap3)
{
  currentTap = tap, cTap2 = tap2, cTap3 = tap3;
}

int64_t FV3_(lfsr)::getTap()
{
  return currentTap;
}

int64_t FV3_(lfsr)::getTap2()
{
  return cTap2;
}

int64_t FV3_(lfsr)::getTap3()
{
  return cTap3;
}

uint64_t FV3_(lfsr)::mls(uint32_t * buffer, uint64_t bufferSize)
{
  return mls32(buffer, bufferSize);
}

uint64_t FV3_(lfsr)::mls32(uint32_t * buffer, uint64_t bufferSize)
{
  long countBit = size_2;
  if(countBit > FV3_MLS_MAX_COUNTBIT) countBit = FV3_MLS_MAX_COUNTBIT;
  uint64_t count = ((uint64_t)1 << countBit)/FV3_MLS_INT_BIT;
  if(count > bufferSize) count = bufferSize;

  memset(buffer, 0, count*sizeof(uint32_t));

  if(size_2 <= 32)
    {
      uint32_t p = currentLFSR, t = currentTap;
      for(uint64_t i = 0;i < count;i ++)
	{
	  for(int b = 0;b < FV3_MLS_INT_BIT;b ++)
	    {
	      // save to each bit
	      buffer[i] = buffer[i] | ((p & 1u) << b);
	      // Galois LFSR
	      p = (p >> 1) ^ (-(int32_t)(p & 1u) & t);
	    }
	}
      currentLFSR = p;
    }
  else if(size_2 <= 64)
    {
      uint64_t p = currentLFSR, t = currentTap;
      for(uint64_t i = 0;i < count;i ++)
	{
	  for(int b = 0;b < FV3_MLS_INT_BIT;b ++)
	    {
	      buffer[i] = buffer[i] | ((p & 1u) << b);
	      p = (p >> 1) ^ (-(int64_t)(p & 1u) & t);
	    }
	}
      currentLFSR = p;
    }
  else if(size_2 <= 128)
    {
      uint64_t p1 = currentLFSR, p2 = cLFSR2, t1 = currentTap, t2 = cTap2;
      for(uint64_t i = 0;i < count;i ++)
	{
	  for(int b = 0;b < FV3_MLS_INT_BIT;b ++)
	    {
	      uint64_t pbit = p1 & 1u;
	      buffer[i] = buffer[i] | (pbit << b);
	      p1 = (p1 >> 1 | p2 << 63) ^ (-(int64_t)pbit & t1);
	      p2 = (p2 >> 1) ^ (-(int64_t)pbit & t2);
	    }
	}
      currentLFSR = p1, cLFSR2 = p2;
    }
  else if(size_2 <= 192)
    {
      uint64_t p1 = currentLFSR, p2 = cLFSR2, p3 = cLFSR3,
	t1 = currentTap, t2 = cTap2, t3 = cTap3;
      for(uint64_t i = 0;i < count;i ++)
	{
	  for(int b = 0;b < FV3_MLS_INT_BIT;b ++)
	    {
	      uint64_t pbit = p1 & 1u;
	      buffer[i] = buffer[i] | (pbit << b);
	      p1 = (p1 >> 1 | p2 << 63) ^ (-(int64_t)pbit & t1);
	      p2 = (p2 >> 1 | p3 << 63) ^ (-(int64_t)pbit & t2);
	      p3 = (p3 >> 1) ^ (-(int64_t)pbit & t3);
	    }
	}
      currentLFSR = p1, cLFSR2 = p2, cLFSR3 = p3;
    }
  else
    {
      throw std::bad_alloc();
    }
  return count;
}


void FV3_(lfsr)::UInt32ToFloat(uint32_t * buffer, fv3_float_t * signal, long bufferSize)
{
  UInt32ToFloat(buffer, signal, bufferSize, 1., -1.);
}

void FV3_(lfsr)::UInt32ToFloat(uint32_t * buffer, fv3_float_t * signal, long bufferSize, fv3_float_t high, fv3_float_t low)
{
  long p = 0;
  for(long j = 0;j < bufferSize;j ++)
    {
      for(int i = 0;i < FV3_MLS_INT_BIT;i ++)
        {
          signal[p] = (buffer[j] >> i) & 0x1 ? high : low;
          p ++;
        }
    }
}

#include "freeverb/fv3_ns_end.h"
